specs/stein/approved/amd-sev-libvirt-support.rst

```diff
@@ -0,0 +1,367 @@
+memory contents, which can be sent to the guest owner as an
+attestation that the memory was encrypted correctly by the firmware.
+Therefore the solution can also provide some degree of protection
+against a malicious hypervisor or compromised cloud operator, although
+it should probably not be relied upon as the sole line of defense.
+
+Use Cases
+---------
+
+#. Booting an instance with a flavor containing an SEV-specific
+   `required trait extra spec
+   <https://docs.openstack.org/nova/latest/user/flavors.html#extra-specs-required-traits>`_
+   will cause that instance to be defined and booted on a compute host
+   which provides that trait.  Additionally the instance should be
+   created with additional XML in the libvirt domain definition which
+   causes QEMU to boot the instance with SEV enabled.
+
+Proposed change
+===============
+
+To achieve the simplest prototype implementation which can possibly
+work, it is proposed that initially during development and testing,
+the placement API would be used to:
+
+- `create a custom trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#trait-create>`_
+  specifically for the purpose of marking flavors as SEV-capable, and
+  then
+
+- `manually assign that trait
+  <https://docs.openstack.org/osc-placement/latest/cli/index.html#resource-provider-trait-set>`_
+  to each compute node which is SEV-capable.
+
+This trait would also indicate that the libvirt driver should include
+extra XML in the guest's domain definition, to enable SEV security,
+ensure that the VM uses machine type ``q35`` and UEFI, and that
+``iommu_platform=on`` is used for all virtio devices, e.g.::
+
+    <domain type='kvm'>
+      <os>
+        <type arch='x86_64' machine='pc-q35-2.11'>hvm</type>
+        <loader readonly='yes' type='pflash'>/usr/share/qemu/ovmf-x86_64-ms-4m-code.bin</loader>
+        <nvram>/var/lib/libvirt/qemu/nvram/sles15-sev-guest_VARS.fd</nvram>
+        <boot dev='hd'/>
+      </os>
+      <launchSecurity type='sev'>
+        <cbitpos>47</cbitpos>
+        <reducedPhysBits>1</reducedPhysBits>
+        <policy>0x0001</policy>
+      </launchSecurity>
+      ...
+    </domain>
+
+For reference, `the AMDSEV GitHub repository
+<https://github.com/AMDESE/AMDSEV/>`_ provides `a complete example of
+SEV-enabled domain XML
+<https://github.com/AMDESE/AMDSEV/blob/master/xmls/sample.xml>`_.
+
+This additional functionality could be implemented within
+``nova.virt.libvirt.LibvirtDriver._set_features()``, e.g. using a new
+``nova.virt.libvirt.LibvirtConfigGuestFeatureSEV`` class.
+
+This work could also mean that images with the property
+``trait:HW_CPU_x86_SEV=required`` would similarly affect the process
+of launching instances.
+
+Immediate follow-up work
+------------------------
+
+It is suggested that follow-up specs be submitted detailing
+the following subsequent work:
+
+- Support a built-in trait which would be automatically detected per
```

Comment:
IMHO this will make the spec approval more complicated because it might not be clear if that is now agreed upon to be implemented or not. if its an optional followup then don't include it in the spec but put it in a separate spec (A new commit in the gerrit sense)

Reply:
OK, you mean submit two commits at once where the second is on top of the first?  That's actually a pretty good idea, but writing a whole second spec is going to delay submission of this even further (I was hoping to submit it today).  How about I just submit it like this for now, and then suggest to upstream that I can push new patch sets (and potentially new commits) which split out the future work into a separate spec?

Reply:
Another option is submitting two specs in a single commit ...